Convert testing to Conda Env instead of Conda Build

Lnaden opened the pull request:
This PR moves the testing environment from relying on Conda-build
and instead creates conda envs which should speed up and simplify tests.

The trade off is the maintenance of multiple files for dependencies
going forward, but I think the benefits outweigh the costs.

Borrows from the QCFractal script from @dgasmith to spin up the
environments. Will need a careful review to make sure it meets our needs.

Fixes #50
Supersedes and Closes #47

Lnaden commented:
cc @jchodera @amjames

Lnaden commented:
> One travis branch can extract the environment from the union of the build and run sections of the meta.yaml file into an environment.yml

There is a fair amount of ambiguity and possible duplicity in trying to do this. The `build`, `host`, and `run` sections of the `meta.yaml` file can often contain items which should appear in one section and not the other (e.g. you don't need all the dependencies to simply install a package, but you do need them to run it), so it runs the risk of creating an env which is not reflective of the shipped product.

Secondly, the Conda Environment file is more flexible than the meta.yaml file since it allow specifying not only channel provenance, but also non-conda sources like Pypi and pip fetching of git repositories. So you could in theory do `meta.yaml -> conda-env`, but the base conda env file you would be porting to would be a partially completed file, and I don't know if we want cookiecutter users to have to figure out all the nuances of conda env files, which are sadly even less documented than meta.yaml files to understand why we only gave them a partial file which we mangle together.

> create an environment, and run the tests; the other can simply run conda build on the recipe

I'd be okay with this, it would just add more tests and I'm not sure we want to overwhelm new user projects with them. I could add a new option to the cookiecutter if you want CI tests of building on top of just testing.

>  we avoid the slowdown of packaging and only then testing

Part of the logic is we are trying to avoid having users worry about deployment out of the gate with this package, and will likely start referring them to things like conda-forge so the user never has to interact with conda-build themselves at all since picking it up is a huge learning curve and gets in the way sometimes. Effectively let someone else handle the deployment (e.g. conda-forge, omnia), you just make sure it works first. This should also be a bit faster since it only has to resolve dependencies once, does not prep a full deployment tarball, then re-resolve and re-install dependencies to install the tarball like conda-build does.

All items for debate though

jchodera commented:
Excellent points! This seems to boil down to the question "what do we want to nudge users toward doing?" Do we want them to just _test_ their code? Or do we want to encourage them to share their code via conda packaging, and make it easy to do so?

I'd _very_ much argue that we want to encourage users to share their code, and that `conda` packages (probably via `conda-forge`) is the best way to do that right now for the CMS community. To me, any marginal improvement in testing speed is not worth the cost of suddenly making it harder to translate this into a conda package and making it very easy to modify the `environment.yml` in a manner that will make it _harder_ to package.

What are _you_ trying to accomplish by this change? Is it _just_ an improvement in travis test times? Is it worth the problems created for packaging?

dgasmith commented:
This cookie cutter aims at getting users setup with best practices and moving in a new repository quickly with low cognitive overhead. `conda-build` is a tricky system (please try explaining host, run, build) that runs counterintuitive to the above. In general, we have a hard time even getting folks to use CI in the first place and throwing up additional barriers is not something that we should do.

I do entirely agree that we want folks on `conda-forge`, which is why we will have a section that when you are ready to release your code, you go to that community to get everything up and running. If you have contributed a package to c-f you know that community is welcoming and open to help, but does have a specific way of writing the `meta.yaml`s that change frequently (and likely will not match up to your local `meta.yaml`).

Translating your `environment.yaml` to a conda-build is at first pass a copy/paste with the c-f community to help you finish the process. I would state that a local `meta.yaml` will have about the same required changes as a copy/paste from `environment.yaml`.

The c-f community specializes in releases, we should continue to move folks in that direction.

Lnaden commented:
Question on the environment creation script for everyone while we work through some of the details.

I use a YAML parser to read and interject a python version before creating the environment. However, there is no native YAML parser for Python. We can have the CI setup install PyYAML or ruamel_yaml (what conda relies on), but that will only be in the base env. The question is should we expect people to have some YAML parser? if so, which one should we expect? 

The alternate is I do some regex parsing and replacement (better than writing a custom yaml-like parser)

dgasmith commented:
Ruamel comes with Conda so no issue there. Is there a reason the produced env needs YAML?

Lnaden commented:
> Is there a reason the produced env needs YAML?

The one case where its an issue is that `conda env` can be invoked from any environment but the act of executing `python conda_create_env ...` from a non-base environment would intuitively work at first glance, but will fail if it can't find a YAML parser unless one exists in that environment.

dgasmith commented:
We could try to use ruamel, poyo, and pyyaml then just throw?

loriab commented:
fwiw, the psi4-dev metapackage with everything from compilers to numpy to testing deps manages to have none of those three yaml pkgs pulled in. Not a barrier, but it does show that you can have a lot installed without yaml.

dgasmith commented:
I wonder if we could turn this into a bash command that uses the conda python which is guaranteed to have ruamel yaml.

Lnaden commented:
Maybe, but I would be hesitant to have a python script try and import modules from a fixed location, that seems very un-pythonic and potentially very risky. I'll instead switch back to using a regex parsing for that one line which needs to be tweaked and drop the need for YAML support at all for now. We can reassess this option in the future if need be

dgasmith commented:
I was mostly thinking that if we use a bash script that finds Conda base python to execute the python script we will have some YAML capabilities via Ruamel. I wouldn’t bother if regex works.

Lnaden commented:
I have changed the script to not rely on YAML at all, it uses pure regex for edits, and its complicated unfortunately to catch all the corner cases people may come up with. I think I got them all, but I left in the commented YAML blocks in case we decide to scrap the regex solution and just tell people to have some YAML module installed. I would appreciate thoughts on the implementation.

Still working on why Windows cannot read the Tempfile, I think I have the solution (a `file.flush()` call), but that might not be it either.

loriab commented on README.md:
```diff
@@ -201,7 +201,22 @@ package managers, Huzzah!
+## Conda Build vs. Conda Environments
+
+We recommend creating Conda environments rather than relying on conda build for *testing* purposes, assuming you have 
+opted for Conda as a dependency manager. Earlier versions of this Cookiecutter would conduct testing with by first 
```
"with by"

loriab commented on README.md:
```diff
@@ -201,7 +201,22 @@ package managers, Huzzah!
+
+The replacement option to this is to pre-create the conda environment and then install your package into it with 
+no dependency resolution for testing. This helps separate out the concepts of **testing** and **deployment** which 
+are separate actions, even though deployment should only come after testing. This should simply and accelerate 
```
simplify

loriab commented on {{cookiecutter.repo_name}}/devtools/scripts/create_conda_env.py:
```diff
@@ -0,0 +1,96 @@
+    if dep_span is not None:
+        final_script += script[dep_span:]
+    script = final_script
+
```
I'm normally regex-happy, but I'll register a slight vote for contortions to get yaml in env rather than the voluminous regex alternative.

Lnaden replied:
I started leaning that way about half way through, but wanted to get a complete answer so everyone could see what it looks like and offer input.

dgasmith replied:
I vote a bit for YAML as well (we could also fall back to regex?). But no real preference as long as it works.

Either way let’s settle on a solution today so that we can run some trials of this Wednesday.

jchodera replied:
Whoa, this looks quite a bit harder to maintain.
What's the problem with YAML?

dgasmith commented:
NamedTempFile will not work on Windows according to the documentation:
```
Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later).
```

It seems you can go through some theatrics [stack](https://stackoverflow.com/questions/15169101/how-to-create-a-temporary-file-that-can-be-read-by-a-subprocess/15235559#15235559):
```
import os, tempfile

DATA = b"hello bob"

def temp_opener(name, flag, mode=0o777):
    return os.open(name, flag | os.O_TEMPORARY, mode)

with tempfile.NamedTemporaryFile() as f:
    f.write(DATA)
    f.flush()
    with open(f.name, "rb", opener=temp_opener) as f:
        assert f.read() == DATA
```

But this would mean changing `conda` I believe which is about as non-starter as it gets. We will need to write, and, to be safe use: `atexit.register(os.unlink, tmpfilename)`.

jchodera commented:
Boo for separating testing and packaging.

dgasmith commented:
If you read through the SO comments that is buggy and will be fixed in 3.8. It seems like we are bending over backwards to use TempFile.

Lnaden commented:
> If you can access Conda you can get that path.

Just because `conda` is on the `PATH` does not mean that `CONDA_EXE` is set right. And the less I have to rely on Env variables for Python scripts, the happier I am

dgasmith commented:
Then grab the python path through Conda info?

Sent from my iPhone

> On Jan 7, 2019, at 5:31 PM, Levi Naden <notifications@github.com> wrote:
> 
> If you can access Conda you can get that path.
> 
> Just because conda is on the PATH does not mean that CONDA_EXE is set right. And the less I have to rely on Env variables for Python scripts, the happier I am
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Lnaden commented:
> It seems like we are bending over backwards to use TempFile.

Fair, I'll come up with something simpler.

Lnaden commented:
> Boo for separating testing and packaging.

But isnt that what we are moving away from? Push people onto Conda-forge or Omnia, or someone else to handle deployment? They should make sure their packages can package up, which is why I think we can add an "allow failure" travis/appveyor builds which only do the build step, thoughts @jchodera ?

Lnaden commented:
* Drop the complicated Regex-Fu in favor of more robust YAML searching  …
* YAML searched by PyYAML->Ruamel YAML->Conda Base Ruamel YAML through `importlib`
  * This is roboust to different conda envs, and different YAML file structures. I can enumerate everything I tried if asked
* Dropped TempFile for TempDirectory which should be more stable on Windows
* Added editing comments from reviewers
* Provided build guidelines to users on what they should chose with conditions
* Added `setup.py sdist` and `conda build` commands to the CI in the "after testing is successful" so builds happen regardless and won't block/slow testing. This ensures builds are still *part of the development process*

This last point I think finds the balance between not slowing down testing, but still including builds as part of the pipeline without the user's thought. Also avoids adding new CI builds and managing the overly exact "Allow Failures" syntax. I imagine though this may be somewhat controversial so I would like some feed back, cc @jchodera @dgasmith @loriab

Lnaden commented:
Once this finishes tests from a squash, I'll merge it
